guest-verona-rt/arch/x86_64/hv_sev.cc

```diff
@@ -0,0 +1,598 @@
+   * Some MSRs are not virtualized and can be written directly with write_msr.
+   * Writing directly is secure, while this function should be viewed as a hint.
+   */
+  static void write_msr_virt_sev(uint32_t msr, uint64_t value)
+  {
+    auto ghcb = get_ghcb();
+
+    ghcb->suffix.format = SevFormat::BASE;
+    ghcb->base.rcx = msr;
+    ghcb->base.rax = value & 0xFFFFFFFF;
+    ghcb->base.rdx = (value >> 32) & 0xFFFFFFFF;
+    ghcb->base.exit_code = SevExitCode::MSR;
+    ghcb->base.exit_info1 = 1;
+    ghcb->base.exit_info2 = 0;
+    ghcb->base.valid_bitmap = SevGhcbValidBitmapData::initial_guest();
+    SEV_GHCB_SET_VALID_BITMAP(ghcb->base.valid_bitmap, rcx);
+    SEV_GHCB_SET_VALID_BITMAP(ghcb->base.valid_bitmap, rax);
+    SEV_GHCB_SET_VALID_BITMAP(ghcb->base.valid_bitmap, rdx);
+
+    vmgexit();
+
+    auto status = ghcb->base.exit_info1;
+    if (status != 0)
+    {
+      LOG_MOD(ERROR, SNP) << "Failed SEV-SNP MSR write with exit code "
+                          << status << "." << LOG_ENDL;
+      kabort();
+    }
+  }
+
+  /**
+   * Write a character to the first COM port using the GHCB protocol.
+   * We don't support other I/O operations as they would trigger more issues.
+   */
+  static void uartputc_sev(uint8_t c)
+  {
+    auto ghcb = get_ghcb();
+
+    ghcb->suffix.format = SevFormat::BASE;
+    ghcb->base.rax = c;
+    ghcb->base.exit_code = SevExitCode::IOIO;
+    ghcb->base.exit_info1 = (1 << 4) | (COM1 << 16);
+    ghcb->base.exit_info2 = 0;
+    ghcb->base.valid_bitmap = SevGhcbValidBitmapData::initial_guest();
+    SEV_GHCB_SET_VALID_BITMAP(ghcb->base.valid_bitmap, rax);
+
+    vmgexit();
+
+    auto status = ghcb->base.exit_info1;
+    if (status != 0)
+    {
+      LOG_MOD(ERROR, SNP) << "Failed SEV OUT with exit code " << status << "."
+                          << LOG_ENDL;
+      kabort();
+    }
+  }
+
+  static void ap_init_sev()
+  {
+    PerCoreData::get()->hypervisor_input_page = setup_ghcb();
+  }
+
+  /**
+   * Heap setup is complex routine under SEV.
+   * Only source of memory information is the loader-injected unmeasured memory
+   * map. The memory map content should be viewed as byzantine. Also need to
+   * check the validity in relation to the known virtual top-of-memory. Abort if
+   * any inconsistency detected.
+   */
+  static void setup_heap_sev(void*)
+  {
+    // Compute the heap start, since the first memory range includes other
+    // elements as well.
+    uint8_t* heap_start = &__heap_start;
+
+    // This cannot be controlled by the hypervisor, but checking just in case
+    // the loader missed it.
+    if (
+      vmsa_settings->virtual_top_of_memory != 0 &&
+      (snmalloc::address_cast(heap_start) >=
+         vmsa_settings->virtual_top_of_memory ||
+       (vmsa_settings->virtual_top_of_memory % PAGE_SIZE) != 0))
+    {
+      kabort();
+    }
+
+    bool first_entry = true;
+    bool first_vtom_entry = true;
+    size_t last_entry_end_address = 0;
```

Comment:
I don't know if you want to do this, but would it be clearer to use `address_t`?
```suggestion
    address_t last_entry_end_address = 0;
```